File: src/animate/sequence.ts
import type { PageInfo } from '@/canvas/canvas.type.js'
import { type Easing, resolveEasing } from '@/animate/easing.js'
import { type Animatable, mix } from '@/animate/interpolate.js'
import { type SpringConfig, assertSpringHasNoRange, spring, springDuration } from '@/animate/spring.js'
import type { Track } from '@/animate/track.js'

/** One leg of a sequence: where to move to, over how long, and how. */
export interface SequenceStep<T extends Animatable> {
  /** Value at the end of this step. The start is wherever the previous step finished. */
  to: T
  /** Seconds this leg lasts. Required unless `spring` supplies one. */
  duration?: number
  /** Easing for this leg. Mutually exclusive with `spring`. @default linear */
  ease?: Easing
  /** Spring physics for this leg instead of an easing. Supplies its own duration. */
  spring?: SpringConfig
  /** Seconds to rest at `to` before the next step begins. @default 0 */
  hold?: number
}

/** A multi-step animation on a single value. */
export interface SequenceConfig<T extends Animatable> {
  /** Value before the sequence starts, and the start of its first step. */
  from: T
  /** Steps, run one after another. */
  steps: SequenceStep<T>[]
  /** Seconds to wait before the first step. @default 0 */
  delay?: number
  /** Extra delay per item index, for offsetting a row of elements. @default 0 */
  stagger?: number
}

/** A step with its timing resolved and its position on the timeline fixed. */
interface PlannedStep<T extends Animatable> {
  from: T
  to: T
  start: number
  duration: number
  ease: ReturnType<typeof resolveEasing>
  spring?: SpringConfig
}

/**
 * Chains several animations on one value, each starting where the last finished.
 *
 * A {@link Track} moves between two values, which covers most of what a scene needs. A value that
 * has to move, wait, and move again — a badge that drops in, pauses, then slides away — would
 * otherwise mean working out each boundary by hand and keeping those numbers in step with every
 * later edit. Declaring the legs instead keeps the arithmetic in one place.
 *
 * Returns the same shape a track does, so the two are interchangeable at the call site.
 * @example
 * ```ts
 * const badge = sequence({
 *   from: -28,
 *   steps: [
 *     { to: 0, spring: { stiffness: 200, damping: 15 } }, // drop in
 *     { to: 0, duration: 0.5, hold: 0.35 },               // rest there
 *     { to: -28, duration: 0.3, ease: 'inCubic' },        // leave
 *   ],
 *   delay: 0.35,
 * })
 *
 * Box({ transform: { translateY: badge.at(page) } })
 * ```
 */
export function sequence<T extends Animatable>(config: SequenceConfig<T>): Track<T> {
  const { from, steps, delay = 0, stagger = 0 } = config

  if (steps.length === 0) {
    throw new Error('[canvas] a sequence needs at least one step')
  }
  if (delay < 0) throw new Error(`[canvas] sequence delay cannot be negative (got ${delay})`)
  if (stagger < 0) throw new Error(`[canvas] sequence stagger cannot be negative (got ${stagger})`)

  const planned: PlannedStep<T>[] = []
  let cursor = delay
  let previous = from

  for (const [index, step] of steps.entries()) {
    if (step.spring && step.ease !== undefined) {
      throw new Error(`[canvas] sequence step ${index} takes \`spring\` or \`ease\`, not both — a spring carries its own curve`)
    }

    if (step.spring) assertSpringHasNoRange(step.spring, `sequence step ${index}`)

    const duration = step.duration ?? (step.spring ? springDuration(step.spring) : undefined)
    if (duration === undefined) {
      throw new Error(`[canvas] sequence step ${index} needs a \`duration\` in seconds, or a \`spring\` to derive one from`)
    }
    if (duration < 0) throw new Error(`[canvas] sequence step ${index} duration cannot be negative (got ${duration})`)

    const hold = step.hold ?? 0
    if (hold < 0) throw new Error(`[canvas] sequence step ${index} hold cannot be negative (got ${hold})`)

    planned.push({
      from: previous,
      to: step.to,
      start: cursor,
      duration,
      ease: resolveEasing(step.ease),
      spring: step.spring,
    })

    // The hold sits after the motion, so the next step begins once the rest is over.
    cursor += duration + hold
    previous = step.to
  }

  const last = planned[planned.length - 1]
  // The trailing hold is not part of the sequence's length: nothing moves during it, so counting it
  // would pad every render that sizes itself from `duration`.
  const end = last.start + last.duration
  const final = last.to

  const at = (page: PageInfo, index = 0): T => {
    const elapsed = page.time - stagger * index

    if (elapsed <= delay) return from
    if (elapsed >= end) return final

    // Linear scan rather than a binary search: a sequence is a handful of steps, and the scan keeps
    // the boundary rules — hold before the next step, motion during a step — in one readable place.
    for (const step of planned) {
      if (elapsed < step.start) return step.from
      if (elapsed >= step.start + step.duration) continue

      const local = elapsed - step.start
      if (step.spring) {
        const unit = spring(local, { ...step.spring, from: 0, to: 1 })
        return mix(step.from, step.to, unit)
      }
      return mix(step.from, step.to, step.ease(step.duration === 0 ? 1 : local / step.duration))
    }

    return final
  }

  return {
    at,
    duration: end,
    totalDuration: (count: number) => end + stagger * Math.max(0, count - 1),
  }
}

File: crates/meo-canvas-core/src/animate/sequence.rs
//! Several motions one after another, and what a group of them lasts.

#![expect(
    clippy::suboptimal_flops,
    reason = "compared bit-for-bit against v1's own numbers; see \
              `animate::easing` for the rule and where it does not apply."
)]

use crate::{
    Error,
    animate::{interpolate::Animatable, track::Motion},
};

/// One leg of a sequence: where to go, how long to take, and how long to wait
/// afterwards.
#[derive(Debug, Clone, Copy, PartialEq)]
pub struct Step<T> {
    /// Where this leg ends. It starts wherever the last one finished.
    pub to: T,
    /// How long the motion takes. `None` asks a spring for its settling time.
    pub duration: Option<f64>,
    /// What carries it.
    pub motion: Motion,
    /// Stillness after the motion, before the next leg begins.
    pub hold: f64,
}

/// A run of steps from one starting value.
#[derive(Debug, Clone, PartialEq)]
pub struct Sequence<T> {
    /// Where the whole run starts.
    pub from: T,
    /// The legs, in order.
    pub steps: Vec<Step<T>>,
    /// Seconds before the first leg.
    pub delay: f64,
    /// Extra delay per index, so a row of things can run in turn.
    pub stagger: f64,
}

/// One leg with its timing worked out.
#[derive(Debug, Clone, Copy, PartialEq)]
struct Leg<T> {
    from: T,
    to: T,
    start: f64,
    duration: f64,
    motion: Motion,
}

/// A sequence whose timing has been resolved and checked.
///
/// **Planned once rather than per sample.** v1 validates while building and
/// throws there, then samples cheaply; the same split here means
/// [`Plan::at`] cannot fail, so a caller sampling a hundred pages checks the
/// arithmetic once rather than a hundred times.
#[derive(Debug, Clone, PartialEq)]
pub struct Plan<T> {
    from: T,
    legs: Vec<Leg<T>>,
    delay: f64,
    stagger: f64,
    end: f64,
}

impl<T: Animatable> Sequence<T> {
    /// Works out when each leg runs, refusing a sequence that does not describe
    /// a motion.
    ///
    /// # Errors
    ///
    /// Returns [`Error::Track`] for an empty sequence, a negative delay,
    /// stagger, duration or hold, or an eased step with no duration -- a curve
    /// has no length of its own where a spring does.
    pub fn plan(&self) -> Result<Plan<T>, Error> {
        if self.steps.is_empty() {
            return Err(Error::Track("sequence needs at least one step"));
        }
        if self.delay < 0.0 {
            return Err(Error::Track("delay cannot be negative"));
        }
        if self.stagger < 0.0 {
            return Err(Error::Track("stagger cannot be negative"));
        }

        let mut legs = Vec::with_capacity(self.steps.len());
        let mut cursor = self.delay;
        let mut previous = self.from;

        for step in &self.steps {
            if step.hold < 0.0 {
                return Err(Error::Track("a step's hold cannot be negative"));
            }
            let duration = match (step.duration, step.motion) {
                (Some(seconds), _) => seconds,
                (None, Motion::Spring(shape)) => {
                    shape.over(0.0, 1.0).settles_after(
                        crate::animate::spring::DEFAULT_REST_DELTA,
                    )?
                }
                (None, Motion::Ease(_)) => {
                    return Err(Error::Track(
                        "an eased step needs a duration in seconds",
                    ));
                }
            };
            if duration < 0.0 {
                return Err(Error::Track(
                    "a step's duration cannot be negative",
                ));
            }

            legs.push(Leg {
                from: previous,
                to: step.to,
                start: cursor,
                duration,
                motion: step.motion,
            });
            // The hold sits after the motion, so the next leg begins once the
            // rest is over.
            cursor += duration + step.hold;
            previous = step.to;
        }

        let last = legs[legs.len() - 1];
        Ok(Plan {
            from: self.from,
            // **The trailing hold is not part of the length**: nothing moves
            // during it, so counting it would pad every render that sizes
            // itself from the duration.
            end: last.start + last.duration,
            legs,
            delay: self.delay,
            stagger: self.stagger,
        })
    }
}

impl<T: Animatable> Plan<T> {
    /// The value at `seconds`, for the `index`th of a staggered set.
    #[must_use]
    pub fn at(&self, seconds: f64, index: usize) -> T {
        #[expect(
            clippy::cast_precision_loss,
            reason = "an index past 2^53 would be a staggered set larger than \
                      the arena can hold"
        )]
        let elapsed = seconds - self.stagger * index as f64;

        if elapsed <= self.delay {
            return self.from;
        }
        let last = self.legs[self.legs.len() - 1];
        if elapsed >= self.end {
            return last.to;
        }

        // A linear scan rather than a search: a sequence is a handful of legs,
        // and the scan keeps both boundary rules -- holding before a leg,
        // moving during one -- in one readable place.
        for leg in &self.legs {
            if elapsed < leg.start {
                return leg.from;
            }
            if elapsed >= leg.start + leg.duration {
                continue;
            }
            let local = elapsed - leg.start;
            return match leg.motion {
                Motion::Spring(shape) => shape
                    .over(0.0, 1.0)
                    .at(local)
                    .map_or(leg.to, |unit| leg.from.mix(leg.to, unit)),
                Motion::Ease(ease) => {
                    let t = if leg.duration == 0.0 {
                        1.0
                    } else {
                        local / leg.duration
                    };
                    leg.from.mix(leg.to, ease.at(t))
                }
            };
        }
        last.to
    }

    /// How long one member of the set runs for.
    #[must_use]
    pub const fn duration(&self) -> f64 {
        self.end
    }

    /// How long a staggered set of `count` of these runs for.
    #[must_use]
    pub const fn total_duration(&self, count: usize) -> f64 {
        #[expect(clippy::cast_precision_loss, reason = "as `Plan::at`")]
        let last = count.saturating_sub(1) as f64;
        self.end + self.stagger * last
    }
}

/// How long a group of motions started together lasts.
///
/// **This is what survives of v1's `parallel`.** There, a group is a record of
/// named members whose `at` returns a record of their values -- and that
/// record is a *type*, assembled by TypeScript's mapped types from whatever
/// was passed in. Rust has no need of it: a caller with three tracks writes a
/// struct with three fields and calls each, which is what the mapped type was
/// reconstructing. **What does not fall out for free is the timing**, because
/// the members have different value types and only their durations are
/// comparable.
///
/// So the group is over when its longest member is. Returns `None` for an
/// empty group, which v1 refuses outright -- **a group of nothing has no
/// duration rather than a duration of zero**, and zero would read as
/// "finished" to every caller that checks.
#[must_use]
pub fn longest(durations: &[f64]) -> Option<f64> {
    durations.iter().copied().reduce(f64::max)
}

#[cfg(test)]
mod tests {
    use super::{Sequence, Step, longest};
    use crate::{
        Error,
        animate::{easing::Easing, spring::Shape, track::Motion},
    };

    /// A step that eases to `to` over a second, holding for `hold` after.
    fn step(to: f64, hold: f64) -> Step<f64> {
        Step {
            to,
            duration: Some(1.0),
            motion: Motion::Ease(Easing::Linear),
            hold,
        }
    }

    fn two_legs() -> Sequence<f64> {
        Sequence {
            from: 0.0,
            steps: vec![step(10.0, 1.0), step(20.0, 0.0)],
            delay: 0.0,
            stagger: 0.0,
        }
    }

    #[test]
    fn a_hold_sits_between_the_legs_and_stills_the_value() {
        let plan = two_legs()
            .plan()
            .unwrap_or_else(|error| unreachable!("{error}"));
        assert!((plan.at(0.5, 0) - 5.0).abs() < f64::EPSILON);
        // The first leg lands at 1s and the second does not start until 2s.
        assert!((plan.at(1.0, 0) - 10.0).abs() < f64::EPSILON);
        assert!((plan.at(1.5, 0) - 10.0).abs() < f64::EPSILON);
        assert!((plan.at(2.5, 0) - 15.0).abs() < f64::EPSILON);
        assert!((plan.at(3.0, 0) - 20.0).abs() < f64::EPSILON);
    }

    #[test]
    fn a_trailing_hold_is_not_part_of_the_length() {
        // Nothing moves during it, so counting it would pad every render that
        // sizes itself from the duration.
        let padded = Sequence {
            steps: vec![step(10.0, 5.0)],
            ..two_legs()
        };
        let plan = padded
            .plan()
            .unwrap_or_else(|error| unreachable!("{error}"));
        assert!((plan.duration() - 1.0).abs() < f64::EPSILON);
    }

    #[test]
    fn a_staggered_set_runs_longer_than_one_of_it() {
        let plan = Sequence {
            stagger: 0.25,
            ..two_legs()
        }
        .plan()
        .unwrap_or_else(|error| unreachable!("{error}"));
        assert!((plan.duration() - 3.0).abs() < f64::EPSILON);
        assert!((plan.total_duration(1) - 3.0).abs() < f64::EPSILON);
        assert!((plan.total_duration(5) - 4.0).abs() < f64::EPSILON);
        // A count of zero is a set with nothing in it rather than a negative
        // stagger applied backwards.
        assert!((plan.total_duration(0) - 3.0).abs() < f64::EPSILON);
    }

    #[test]
    fn delay_holds_the_start_value_and_shifts_everything() {
        let plan = Sequence {
            delay: 2.0,
            ..two_legs()
        }
        .plan()
        .unwrap_or_else(|error| unreachable!("{error}"));
        assert!((plan.at(1.9, 0)).abs() < f64::EPSILON);
        assert!((plan.at(2.5, 0) - 5.0).abs() < f64::EPSILON);
        assert!((plan.duration() - 5.0).abs() < f64::EPSILON);
    }

    #[test]
    fn a_spring_leg_takes_its_length_from_the_physics() {
        let plan = Sequence {
            steps: vec![Step {
                to: 10.0,
                duration: None,
                motion: Motion::Spring(Shape::default()),
                hold: 0.0,
            }],
            ..two_legs()
        }
        .plan()
        .unwrap_or_else(|error| unreachable!("{error}"));
        assert!(plan.duration() > 0.0);
        assert!(
            (plan.at(plan.duration() + 1.0, 0) - 10.0).abs() < f64::EPSILON
        );
    }

    #[test]
    fn a_zero_duration_leg_lands_rather_than_dividing_by_nothing() {
        let plan = Sequence {
            steps: vec![
                Step {
                    duration: Some(0.0),
                    ..step(10.0, 1.0)
                },
                step(20.0, 0.0),
            ],
            ..two_legs()
        }
        .plan()
        .unwrap_or_else(|error| unreachable!("{error}"));
        assert!(
            plan.at(0.0, 0).is_finite(),
            "a zero-length leg divided by zero"
        );
        assert!((plan.at(1.5, 0) - 15.0).abs() < f64::EPSILON);
    }

    #[test]
    fn a_sequence_that_is_not_a_motion_is_refused() {
        let empty = Sequence {
            steps: Vec::new(),
            ..two_legs()
        };
        assert!(matches!(empty.plan(), Err(Error::Track(_))));
        assert!(matches!(
            Sequence {
                delay: -1.0,
                ..two_legs()
            }
            .plan(),
            Err(Error::Track(_))
        ));
        assert!(matches!(
            Sequence {
                stagger: -1.0,
                ..two_legs()
            }
            .plan(),
            Err(Error::Track(_))
        ));
        assert!(matches!(
            Sequence {
                steps: vec![step(1.0, -1.0)],
                ..two_legs()
            }
            .plan(),
            Err(Error::Track(_))
        ));
        assert!(matches!(
            Sequence {
                steps: vec![Step {
                    duration: Some(-1.0),
                    ..step(1.0, 0.0)
                }],
                ..two_legs()
            }
            .plan(),
            Err(Error::Track(_))
        ));
        assert!(matches!(
            Sequence {
                steps: vec![Step {
                    duration: None,
                    ..step(1.0, 0.0)
                }],
                ..two_legs()
            }
            .plan(),
            Err(Error::Track(_))
        ));
    }

    #[test]
    fn a_group_of_nothing_has_no_duration() {
        // Not zero: zero reads as finished to every caller that checks.
        assert!(longest(&[]).is_none());
        assert!(
            (longest(&[1.0, 3.5, 2.0]).unwrap_or_else(|| unreachable!(
                "a non-empty group has a duration"
            )) - 3.5)
                .abs()
                < f64::EPSILON
        );
    }
}
